client/me/purchases/payment-methods/payment-method-edit.tsx

```diff
@@ -0,0 +1,158 @@
+import { Button } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import { FunctionComponent, useState, useCallback } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import {
+	isPaymentAgreement,
+	getPaymentMethodSummary,
+	PaymentMethod,
+} from 'calypso/lib/checkout/payment-methods';
+import { errorNotice, successNotice } from 'calypso/state/notices/actions';
+import { ReduxDispatch } from 'calypso/state/redux-store';
+import { editStoredCardTaxLocation } from 'calypso/state/stored-cards/actions';
+import { isEditingStoredCard } from 'calypso/state/stored-cards/selectors';
+import PaymentMethodDetails from './payment-method-details';
+import PaymentMethodEditDialog from './payment-method-edit-dialog';
+
+interface Props {
+	card: PaymentMethod;
+}
+
+const PaymentMethodEdit: FunctionComponent< Props > = ( { card } ) => {
+	const translate = useTranslate();
+	const isEditing = useSelector( ( state ) =>
+		isEditingStoredCard( state, card.stored_details_id )
+	);
+
+	const reduxDispatch = useDispatch< ReduxDispatch >();
+	const [ isDialogVisible, setIsDialogVisible ] = useState( false );
+	const closeDialog = useCallback( () => setIsDialogVisible( false ), [] );
+
+	const renderTaxPostalCode = (): string => {
+		const filtered = card.meta.find(
+			( item: { meta_key: string } ) => item.meta_key === 'tax_postal_code'
+		);
+		return filtered?.meta_value ?? '';
+	};
+
+	const renderTaxCountryCode = (): string => {
+		const filtered = card.meta.find(
+			( item: { meta_key: string } ) => item.meta_key === 'tax_country_code'
+		);
+		return filtered?.meta_value ?? '';
+	};
+
+	const [ inputValues, setInputValues ] = useState( {
+		tax_postal_code: '',
+		tax_country_code: '',
+	} );
+
+	const handleEdit = useCallback( () => {
+		closeDialog();
+
+		reduxDispatch(
+			editStoredCardTaxLocation( card, inputValues.tax_postal_code, inputValues.tax_country_code )
+		)
+			.then( () => {
+				if ( isPaymentAgreement( card ) ) {
+					reduxDispatch( successNotice( translate( 'Payment method edited successfully' ) ) );
+				} else {
+					reduxDispatch( successNotice( translate( 'Card edited successfully!' ) ) );
+				}
+				recordTracksEvent( 'calypso_purchases_edit_tax_location' );
+			} )
+			.then( () => {
+				window.location.reload();
+			} )
+			.catch( ( error: Error ) => {
+				reduxDispatch( errorNotice( error.message ) );
+			} );
+	}, [
+		closeDialog,
+		card,
+		inputValues.tax_postal_code,
+		inputValues.tax_country_code,
+		translate,
+		reduxDispatch,
+	] );
+
+	const handleSubmit = ( event ) => {
+		event.preventDefault();
+		const { name, value } = event.target;
+		setInputValues( { ...inputValues, [ name ]: value } );
+		handleEdit();
```

Comment:
The previous line requests saving the new value, but it doesn't guarantee that it will be updated by the time `handleEdit` is run. Note how `useState` works: it returns two values, the current value of the data and a setter. When you call the setter, the current value is not changed immediately; you don't get the new value until `useState` is run again, which happens on the next render of the component. Calling the setter does queue an update of the component, so it will happen very soon, but not soon enough for `handleEdit` here.

You can solve this various ways. One way is to use the stored tax values directly as the source of data for the form instead of having your own state; that way, whenever the values are updated by your redux-dispatched action, the component will re-render with the new values. Another way would be to pass the new values as arguments directly to `handleEdit` since you have them here, rather than relying on getting them from local state.